anndata/readwrite/h5ad.py

```diff
@@ -178,9 +179,15 @@ def write_series(f, key, series, dataset_kwargs={}):
             **dataset_kwargs,
         )
     elif is_categorical_dtype(series):
-        f.create_dataset(key, shape=series.shape, dtype=series.cat.codes.dtype)
-        f[key][:] = series.cat.codes
-        f[key].attrs["categories"] = list(series.cat.categories)
+        cats = series.cat.categories.values
+        codes = series.cat.codes.values
+        category_key = f"_{key}_categories"
+        if category_key in f:
+            # TODO: figure out what to do in case of name collision
+            raise NotImplementedError()
```

Comment:
What are the options?

Reply:
Not sure yet. Goals are:

1. No restrictions on column names for the user
2. Human readable

What I'm at is close, and will probably extend with just appending things. On the other hand I could ditch goal 2 and go with a uuid.

Reply:
Here's a draft fix. It's not my favorite, but it'll work. I'm hoping there's a more clever solution

<details>
<summary>Possible fix </summary>

```python
def write_dataframe(f, key, df, dataset_kwargs={}):
    group = f.h5f.create_group(key)
    group.attrs["encoding-type"] = "dataframe"
    group.attrs["encoding-version"] = "0.1.0"
    group.attrs["column-order"] = list(df.columns)

    if df.index.name is not None:
        index_name = df.index.name
    else:
        index_name = "_index"
    group.attrs["_index"] = index_name

    reserved_names = [index_name] + list(df.columns)

    write_series(group, index_name, df.index, reserved_names, dataset_kwargs)
    for colname, series in df.items():
        write_series(group, colname, series, reserved_names, dataset_kwargs)


def write_series(f, key, series, reserved_names=(), dataset_kwargs={}):
    # f has to be actual h5py type, otherwise categoricals won't write
    if series.dtype == object:  # Assuming it's string
        f.create_dataset(
            key,
            data=series.values,
            dtype=h5py.special_dtype(vlen=str),
            **dataset_kwargs,
        )
    elif is_categorical_dtype(series):
        cats = series.cat.categories.values
        codes = series.cat.codes.values
        category_key = f"_{key}_categories"

        # Permute to get unique value
        cur = -1
        while category_key in f or category_key in reserved_names:
            category_key = category_key.rstrip(str(cur))
            cur += 1
            category_key += str(cur)

        write_array(f, category_key, cats, dataset_kwargs)
        code_dset = f.create_dataset(key, data=codes)
        code_dset.attrs["categories"] = f[category_key].ref
    else:
        f[key] = series.values
```

</details>

Reply:
You can’t set the array of categories directly as attribute of the dataset containing the `codes`, right? You have to use the `ref`?

Reply:
> How about storing them in a group called categories?

I like that one. I think I'll go with `__categories__` and make it a reserved value.

We should be able to tell if its a sparse dataset from it's attributes, and we also probably won't support storing a sparse matrix as a column of a data frame in the near future.

Reply:
I'm thinking I should copy this over to zarr